Sources/NIOEmbedded/AsyncEmbeddedEventLoop.swift

```diff
@@ -0,0 +1,357 @@
+            }
+        }, promise.fail)
+
+        self.scheduledTasks.push(task)
+    }
+
+    /// - see: `EventLoop.scheduleTask(deadline:_:)`
+    @discardableResult
+    public func scheduleTask<T>(deadline: NIODeadline, _ task: @escaping () throws -> T) -> Scheduled<T> {
+        let promise: EventLoopPromise<T> = self.makePromise()
+        let taskID = self.scheduledTaskCounter.add(1)
+
+        let scheduled = Scheduled(promise: promise, cancellationTask: {
+            if self.inEventLoop {
+                self.removeTask(taskID: taskID)
+            } else {
+                self.queue.async {
+                    self.removeTask(taskID: taskID)
+                }
+            }
+        })
+
+        if self.inEventLoop {
+            self.insertTask(taskID: taskID, deadline: deadline, promise: promise, task: task)
+        } else {
+            self.queue.async {
+                self.insertTask(taskID: taskID, deadline: deadline, promise: promise, task: task)
+            }
+        }
+        return scheduled
+    }
+
+    /// - see: `EventLoop.scheduleTask(in:_:)`
+    @discardableResult
+    public func scheduleTask<T>(in: TimeAmount, _ task: @escaping () throws -> T) -> Scheduled<T> {
+        return self.scheduleTask(deadline: self.now + `in`, task)
+    }
+
+    /// On an `EmbeddedEventLoop`, `execute` will simply use `scheduleTask` with a deadline of _now_. This means that
+    /// `task` will be run the next time you call `EmbeddedEventLoop.run`.
+    public func execute(_ task: @escaping () -> Void) {
+        self.scheduleTask(deadline: self.now, task)
+    }
+
+    /// Run all tasks that have previously been submitted to this `EmbeddedEventLoop`, either by calling `execute` or
+    /// events that have been enqueued using `scheduleTask`/`scheduleRepeatedTask`/`scheduleRepeatedAsyncTask` and whose
+    /// deadlines have expired.
+    ///
+    /// - seealso: `EmbeddedEventLoop.advanceTime`.
+    public func run() async {
+        // Execute all tasks that are currently enqueued to be executed *now*.
+        await self.advanceTime(to: self.now)
+    }
+
+    /// Runs the event loop and moves "time" forward by the given amount, running any scheduled
+    /// tasks that need to be run.
+    public func advanceTime(by increment: TimeAmount) async {
+        await self.advanceTime(to: self.now + increment)
+    }
+
+    /// Unwrap a future result from this event loop.
+    ///
+    /// This replaces `EventLoopFuture.get()` for use with `AsyncEmbeddedEventLoop`. This is necessary because attaching
+    /// a callback to an `EventLoopFuture` (which is what `EventLoopFuture.get` does) requires scheduling a work item onto
+    /// the event loop, and running that work item requires spinning the loop. This is a non-trivial dance to get right due
+    /// to timing issues, so we need to
+    public func awaitFuture<ResultType: Sendable>(_ future: EventLoopFuture<ResultType>) async throws -> ResultType {
```

Comment:
I don't quite like the name because on the call site you will write `await` twice e.g.
```swift
let result = try await el.awaitFuture(someFuture)
```
I'm think about moving this into an extension on `EventLoopFuture` like this:
```swift
extension EventLoopFuture {
    func get(on eventLoop: AsyncEmbeddedEventLoop) async throws -> Value {
        ...
    }
}
let result = try await someFuture.get(on: el)
```
This way it would also show up in autocomplete when someone tries to call the "wrong" `.get()`.

Reply:
> I don't quite like the name because on the call site you will write await twice

I agree.

> This way it would also show up in autocomplete when someone tries to call the "wrong" .get().

However there is no way for the user to know which is the right and which is the wrong `get()`. Why would I sometimes supply the EventLoop, if it is optional.

Maybe:

```swift
let result = try await el.runUntil(someFuture)
```

Reply:
Hmm, I don't love `runUntil` either. `unwrap`?